src/transactions/ClaimClaimableBalanceOpFrame.cpp

```diff
@@ -0,0 +1,164 @@
+        return pred.absBefore() > closeTime;
+    case CLAIM_PREDICATE_AFTER_ABSOLUTE_TIME:
+        return pred.absAfter() < closeTime;
+    case CLAIM_PREDICATE_BEFORE_RELATIVE_TIME:
+    case CLAIM_PREDICATE_AFTER_RELATIVE_TIME:
+        throw std::runtime_error("Invalid ClaimPredicate");
+    default:
+        abort();
+    }
+
+    return true;
+}
+
+bool
+ClaimClaimableBalanceOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    auto ltxEntry =
+        stellar::loadClaimableBalance(ltx, mClaimClaimableBalance.balanceID);
+    if (!ltxEntry)
+    {
+        innerResult().code(CLAIM_CLAIMABLE_BALANCE_DOES_NOT_EXIST);
+        return false;
+    }
+
+    auto const& claimableBalance = ltxEntry.current().data.claimableBalance();
+
+    auto it = std::find_if(
+        std::begin(claimableBalance.claimants),
+        std::end(claimableBalance.claimants), [&](Claimant const& claimant) {
+            return claimant.v0().destination == getSourceID();
+        });
+
+    auto header = ltx.loadHeader();
+    if (it == claimableBalance.claimants.end() ||
+        !validatePredicate(it->v0().predicate,
+                           header.current().scpValue.closeTime))
+    {
+        innerResult().code(CLAIM_CLAIMABLE_BALANCE_CANNOT_CLAIM);
+        return false;
+    }
+
+    // Try to give reserve back to createdBy account. If we can't, try to give
+    // it back to sourceAccount
+    if (!loadAccountAndAddBalance(ltx, header, claimableBalance.createdBy,
+                                  claimableBalance.reserve))
+    {
+        if (!loadAccountAndAddBalance(ltx, header, getSourceID(),
+                                      claimableBalance.reserve))
+        {
+            innerResult().code(CLAIM_CLAIMABLE_BALANCE_LINE_FULL);
+            return false;
+        }
+    }
+
+    auto const& asset = claimableBalance.asset;
+    auto amount = claimableBalance.amount;
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        if (!loadAccountAndAddBalance(ltx, header, getSourceID(), amount))
+        {
+            innerResult().code(CLAIM_CLAIMABLE_BALANCE_LINE_FULL);
+            return false;
+        }
+    }
+    else
+    {
+        auto trustline = loadTrustLine(ltx, getSourceID(), asset);
+        if (!trustline)
+        {
+            innerResult().code(CLAIM_CLAIMABLE_BALANCE_NO_TRUST);
+            return false;
+        }
+        if (!trustline.isAuthorized())
+        {
+            innerResult().code(CLAIM_CLAIMABLE_BALANCE_NOT_AUTHORIZED);
+            return false;
+        }
+        if (!trustline.addBalance(header, amount))
+        {
+            innerResult().code(CLAIM_CLAIMABLE_BALANCE_LINE_FULL);
+            return false;
+        }
+    }
+
+    // remove claimable balance entry
+    ltxEntry.erase();
+
+    innerResult().code(CLAIM_CLAIMABLE_BALANCE_SUCCESS);
+    return true;
+}
+
+bool
+ClaimClaimableBalanceOpFrame::doCheckValid(uint32_t ledgerVersion)
+{
+    return true;
+}
+
```

Comment:
shouldn't there be an override for `insertLedgerKeysToPrefetch` that loads the claimable entry?

Reply:
added the override